File: src/algorithms/arrays/validate-subsequence/solution1.go
package main

// O(n) time | O(1) space
func IsValidSubsequence(array []int, sequence []int) bool {
	arrIdx := 0
	seqIdx := 0
	for arrIdx < len(array) && seqIdx < len(sequence) {
		if array[arrIdx] == sequence[seqIdx] {
			seqIdx += 1
		}
		arrIdx += 1
	}
	return seqIdx == len(sequence)
}

File: src/algorithms/arrays/validate-subsequence/solution1.ts
// O(n) time | O(1) space
export function isValidSubsequence(array: number[], sequence: number[]) {
  let arrIdx = 0
  let seqIdx = 0
  while (arrIdx < array.length && seqIdx < sequence.length) {
    if (array[arrIdx] === sequence[seqIdx]) seqIdx++
    arrIdx++
  }
  return seqIdx === sequence.length
}
